rfc/custom-modules-v1.md

```diff
@@ -0,0 +1,706 @@
+	Response *core.GraphQLResponse
+	// The original http response from the subgraph
+	HttpResponse *http.Response
+}
+
+// Router Hooks
+
+type RouterRequestHook interface {
+	// OnRouterRequest is called when a request is made to the router and after all GraphQL information is available
+	// Returning an error will result in a GraphQL error being returned to the client.
+	OnRouterRequest(ctx *core.RouterRequest, err error) error
+}
+
+type RouterResponseHook interface {
+	// OnRouterResponse is called before the response is sent to the client
+	// Returning an error will result in a GraphQL error being returned to the client.
+	OnRouterResponse(ctx *core.RouterResponse, err error) error
+}
+
+type RouterErrorHook interface {
+	// OnError is called when an error occurs during the router lifecycle
+	OnRouterError(err error)
+}
+
+// SubgraphHooks are called when a subgraph request or response is made.
+// The order is not guaranteed, so the hooks should be idempotent and side-effect free.
+// if state needs to be shared between hooks, it should be stored in the context.
+// We will provide an easy way to share state between hooks.
+
+type SubgraphRequestHook interface {
+	// OnSubgraphRequest is called when a subgraph request is made
+	// Returning an error will result in a GraphQL error being returned to the client.
+	OnSubgraphRequest(ctx *core.SubgraphRequest, err error) error
+}
+
+type SubgraphResponseHook interface {
+	// OnSubgraphResponse is called when a subgraph response is received
+	// Returning an error will result in a GraphQL error being returned to the client.
+	OnSubgraphResponse(ctx *core.SubgraphResponse, err error) error
+}
+
+type SubgraphErrorHook interface {
+	// OnError is called when an error occurs during the subgraph lifecycle
+	OnSubgraphError(err error)
+}
+
+// ApplicationHooks are called when the application starts, stops, or an error occurs.
+
+type ApplicationStartHook interface {
+	// OnAppStart is called when the application starts
+	// Returning an error will result in the application not starting
+	OnAppStart() error
+}
+
+type ApplicationStopHook interface {
+	// OnAppStop is called when the application stops
+	// Returning an error will result in the application not stopping
+	OnAppStop() error
+}
+
+type ApplicationErrorHook interface {
+	// OnError is called when an error occurs during the application lifecycle start/stop or any other error
+	OnAppError(err error)
+}
+
+// TelemetryHooks are called when a span is created
+
+type TelemetrySpanHook interface {
+	// OnSpan is called when a span is created
+	// Returning a function to be called when the span ends.
+	// This can be used to add custom attributes or events to the span.
+	OnSpan(span *trace.Span) func() // Return a function to be called when the span ends
+}
+
+type TelemetryMetricHook interface {
+	// OnMetric is called when a metric is recorded
+	// Returning an error will result in a telemetry error
+	OnMetric(metric *metric.Metric) error
+}
```

Comment:
I see OnSpan() and OnMetric() can be useful in some cases. But adding attributes can be achieved by adding keyValue pairs in request context (please correct me if I'm wrong).

This is a different level of abstraction. I would strongly prefer a more flexible hook system, eg, using Pre/Post, that allows to metric/tracing plugins. We can still provide default Prometheus, OTel metrics/tracings by building default plugins on top of the hook system.